src/Utilities/Registration.hpp

```diff
@@ -0,0 +1,54 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup UtilitiesGroup
+/// \brief Helpers for derived class registration
+///
+/// SpECTRE's factory mechanism requires that the base class contain a
+/// list of all derived classes.  This can be problematic when the
+/// desired list of derived classes can vary with use of the base
+/// class.  These helpers provide a method to simplify handling these
+/// compile-time registrations.
+///
+/// Each optional derived class defines a registrar helper and both
+/// the base class and derived classes are templated on the list of
+/// registrars.  The typical structure of these classes is:
+/// \snippet Test_Registration.cpp registrar_structure
+/// A concrete base class type with a specific set of derived classes
+/// is constructed like
+/// \snippet Test_Registration.cpp registrar_use
+namespace Registration {
+/// A template for defining a registrar.
+///
+/// A registrar for a class can be defined by making a type alias to
+/// this struct, filling in the registrant.
+/// \snippet Test_Registration.cpp registrar
+/// In more complex cases (such as with non-type template
+/// parameters) defining a registrar manually may be necessary.
+/// \snippet Test_Registration.cpp custom_registrar
+template <template <typename...> class Registrant, typename... Args>
+struct Registrar {
+  template <typename Registrars>
+  using f = Registrant<Args..., Registrars>;
```

Comment:
`f` is a standard name we use (similar `type`) but denotes template nested aliases. This naming convention is what's used in continuation-based metaprogramming.

Reply:
Do you have a suggestion?  The only thing I'm coming up with is "Registrant specialized to RegistrarList", but I think that's less clear than the code it is commenting on.

Reply:
When we were reviewing this, the biggest hurdle I found to understanding the code was working out that this was the point where the potentially incomplete type (i.e. still a template) `Registrant` ended up getting specialized to a definite type via the other template parameters. Perhaps something similar to: "final Registrant type, specialized with input `Args...` and the `RegistrarList` it was extracted from. This will be a full specialization of a derived class which uses the registration framework." (unless, of course, that is a misunderstanding of what's going on...)?

Reply:
OK, done.  (By the way, you should look up "(in)complete type" because it does not mean what you are using it to mean and I was confused by this.)

Reply:
oops -- thanks for the correction. I guess the correct way of phrasing this is 'a class template rather than a class specialization'?

Reply:
Yeah.  Maybe "class template specialization".  I forget the exact term there, but I don't think that one would lead to confusion either way.